datastore/src/main/scala/org/allenai/datastore/Datastore.scala

```diff
@@ -162,6 +80,16 @@ class Datastore(val name: String, val s3: AmazonS3Client) extends Logging {
     cause
   )
 
+  /** Exception indicating that the locator provided was for the wrong datastore. This is thrown if
+    * the "datastore" field in a locator doesn't match this datastore's name.
+    * @param locator the bad locator
+    */
+  class BadLocatorException(locator: Locator)
+    extends DsException(
+      s"locator for ${locator.datastore} doesn't match this datastore ($name)",
+      null
+    )
```

Comment:
That's true, and it's not super sexy. I guess it depends on the use cases. I think that most of the time, we create an instance of `Datastore` to get exactly one file. For those use cases, this new approach is better. But there are probable cases where we get a whole bunch of files, and then we make it harder.

Overall, without digging into all the cases, I'd say that this idea of version 2.0 is still better.